File: go/processor/dsti/naming_analyzer.go
// @docspec:module {
//   id: "docspec-go-dsti-naming",
//   name: "DSTI Naming Analyzer",
//   description: "Parses function and method names into verb-object-intent triples using camelCase decomposition, mirroring the Java NamingAnalyzer for DSTI channel 1 (nameSemantics).",
//   since: "3.0.0"
// }
package dsti

import (
	"strings"
	"unicode"
)

// NameSemantics holds the parsed semantic components of a function name.
// @docspec:intentional "Captures the verb-object-intent triple from a function name"
type NameSemantics struct {
	Verb   string `json:"verb"`
	Object string `json:"object"`
	Intent string `json:"intent"`
}

// Known verb prefixes ordered longest-first where needed to avoid prefix collisions.
var verbPrefixes = []string{
	"get", "set", "is", "has", "find", "create", "delete", "remove",
	"update", "save", "add", "validate", "check", "compute", "calculate",
	"process", "handle", "convert", "transform", "parse", "format",
	"build", "generate", "initialize", "init", "load", "fetch",
	"send", "receive", "publish", "subscribe", "notify", "dispatch",
	"sync", "migrate", "schedule", "retry", "batch",
	"aggregate", "merge", "split", "emit", "enrich", "filter",
	"list", "count", "new",
}

// intentMap maps verb prefixes to intent categories, matching the Java
// NamingAnalyzer's categorizeIntent switch statement.
var intentMap = map[string]string{
	"get":        "query",
	"find":       "query",
	"fetch":      "query",
	"load":       "query",
	"is":         "query",
	"has":        "query",
	"list":       "query",
	"count":      "query",
	"set":        "mutation",
	"update":     "mutation",
	"save":       "mutation",
	"add":        "mutation",
	"create":     "creation",
	"build":      "creation",
	"generate":   "creation",
	"initialize": "creation",
	"init":       "creation",
	"new":        "creation",
	"delete":     "deletion",
	"remove":     "deletion",
	"validate":   "validation",
	"check":      "validation",
	"compute":    "computation",
	"calculate":  "computation",
	"process":    "computation",
	"convert":    "transformation",
	"parse":      "transformation",
	"format":     "transformation",
	"transform":  "transform",
	"handle":     "handler",
	"dispatch":   "handler",
	"send":       "emission",
	"receive":    "consumption",
	"sync":       "synchronize",
	"migrate":    "migrate",
	"schedule":   "schedule",
	"retry":      "retry",
	"batch":      "batch-process",
	"aggregate":  "aggregate",
	"merge":      "merge",
	"split":      "split",
	"notify":     "notify",
	"emit":       "emit",
	"publish":    "publish",
	"subscribe":  "subscribe",
	"enrich":     "enrich",
	"filter":     "filter",
}

// AnalyzeName parses a function or method name and extracts its semantic
// components: verb, object, and intent category. This mirrors the Java
// NamingAnalyzer.analyze() method.
//
// @docspec:method { since: "3.0.0" }
// @docspec:deterministic
// @docspec:intentional "Decomposes a function name into verb-object-intent using camelCase boundaries and a known verb prefix table"
//
// Examples:
//
//	"GetUserByID" -> {Verb: "get", Object: "user by id", Intent: "query"}
//	"CreateOrder" -> {Verb: "create", Object: "order", Intent: "creation"}
//	"ValidateInput" -> {Verb: "validate", Object: "input", Intent: "validation"}
func AnalyzeName(name string) NameSemantics {
	lower := strings.ToLower(name)

	for _, verb := range verbPrefixes {
		if !strings.HasPrefix(lower, verb) || len(name) <= len(verb) {
			continue
		}

		// Verify the next character after the verb is uppercase (camelCase boundary)
		nextChar := rune(name[len(verb)])
		if !unicode.IsUpper(nextChar) {
			continue
		}

		object := name[len(verb):]
		intent := intentMap[verb]
		if intent == "" {
			intent = "unknown"
		}

		return NameSemantics{
			Verb:   verb,
			Object: splitCamelCaseToLower(object),
			Intent: intent,
		}
	}

	// Fallback: the entire name is the verb
	return NameSemantics{
		Verb:   strings.ToLower(name),
		Object: "",
		Intent: "unknown",
	}
}

// splitCamelCaseToLower splits a camelCase string into space-separated lowercase words.
// Example: "UserProfile" -> "user profile"
//
// @docspec:deterministic
func splitCamelCaseToLower(s string) string {
	words := SplitCamelCase(s)
	for i, w := range words {
		words[i] = strings.ToLower(w)
	}
	return strings.Join(words, " ")
}

// SplitCamelCase splits a camelCase or PascalCase string into its component words.
// Handles runs of uppercase letters (acronyms) correctly:
//
// @docspec:method { since: "3.0.0" }
// @docspec:deterministic
//
//	"HTTPServer" -> ["HTTP", "Server"]
//	"getUserByID" -> ["get", "User", "By", "ID"]
func SplitCamelCase(s string) []string {
	if s == "" {
		return nil
	}

	var words []string
	var current strings.Builder

	runes := []rune(s)
	for i := 0; i < len(runes); i++ {
		r := runes[i]

		if i == 0 {
			current.WriteRune(r)
			continue
		}

		if unicode.IsUpper(r) {
			// Check if this is part of an acronym (next char is also upper or end of string)
			if unicode.IsUpper(runes[i-1]) {
				// Could be mid-acronym: check if next is lowercase (end of acronym)
				if i+1 < len(runes) && unicode.IsLower(runes[i+1]) {
					// End of acronym, start new word
					if current.Len() > 0 {
						words = append(words, current.String())
						current.Reset()
					}
				}
				current.WriteRune(r)
			} else {
				// Normal camelCase boundary
				if current.Len() > 0 {
					words = append(words, current.String())
					current.Reset()
				}
				current.WriteRune(r)
			}
		} else {
			current.WriteRune(r)
		}
	}

	if current.Len() > 0 {
		words = append(words, current.String())
	}

	return words
}

File: rust/processor/src/dsti/naming_analyzer.rs
//! Parses Rust function/method names to extract semantic verbs and objects.
//!
//! @docspec:module {
//!   id: "docspec-rust-dsti-naming",
//!   name: "Naming Analyzer",
//!   description: "Extracts semantic verb-object pairs and intent categories from snake_case Rust function names, mirroring the Java NamingAnalyzer for camelCase.",
//!   since: "3.0.0"
//! }
//!
//! Uses snake_case conventions (unlike the Java NamingAnalyzer which uses camelCase).
//! Used by the DSTI (Documentation Semantic & Temporal Intelligence) system to infer
//! developer intent from naming patterns.
//!
//! Mirrors Java `io.docspec.processor.dsti.NamingAnalyzer`.

/// Semantic components extracted from a function name.
///
/// @docspec:boundary "Carries the decomposed verb, object, and intent from a single function name"
#[derive(Debug, Clone, PartialEq)]
pub struct NameSemantics {
    /// The verb prefix (e.g., "get", "create", "validate").
    pub verb: String,
    /// The object/noun portion (e.g., "user", "payment order").
    pub object: String,
    /// The inferred intent category (e.g., "query", "mutation", "creation").
    pub intent: String,
}

/// Known verb prefixes, ordered longest-first where needed to avoid prefix collisions.
const VERBS: &[&str] = &[
    "get", "set", "is", "has", "find", "create", "delete", "remove",
    "update", "save", "add", "validate", "check", "compute", "calculate",
    "process", "handle", "convert", "transform", "parse", "format",
    "build", "generate", "initialize", "init", "load", "fetch",
    "send", "receive", "publish", "subscribe", "notify", "dispatch",
    "sync", "migrate", "schedule", "retry", "batch",
    "aggregate", "merge", "split", "emit", "enrich", "filter",
    "list", "count", "verify", "execute", "run", "start", "stop",
    "open", "close", "read", "write", "flush", "drain",
    "encode", "decode", "serialize", "deserialize",
    "connect", "disconnect", "bind", "unbind",
    "register", "unregister", "mount", "unmount",
    "lock", "unlock", "acquire", "release",
    "try", "poll", "await", "spawn",
];

/// Analyze a function name (snake_case) and extract its semantic components.
///
/// @docspec:deterministic
/// @docspec:intentional "Decomposes a snake_case function name into verb, object, and intent category using a known verb dictionary"
///
/// # Examples
/// ```ignore
/// let result = analyze("get_user_by_id");
/// assert_eq!(result.verb, "get");
/// assert_eq!(result.object, "user by id");
/// assert_eq!(result.intent, "query");
/// ```
pub fn analyze(name: &str) -> NameSemantics {
    // Skip names starting with underscore (private/unused convention)
    let name = name.strip_prefix('_').unwrap_or(name);

    let words: Vec<&str> = name.split('_').collect();
    if words.is_empty() {
        return NameSemantics {
            verb: name.to_string(),
            object: String::new(),
            intent: "unknown".to_string(),
        };
    }

    // Try single-word verb match first
    let first = words[0];
    if VERBS.contains(&first) {
        let object = words[1..].join(" ");
        let intent = categorize_intent(first);
        return NameSemantics {
            verb: first.to_string(),
            object,
            intent,
        };
    }

    // Try two-word verb match (e.g., "try_get" -> verb="try_get")
    if words.len() >= 2 {
        let two_word = format!("{}_{}", words[0], words[1]);
        // Some compound verbs
        let compound_verbs = [
            ("try_get", "query"),
            ("try_send", "emission"),
            ("try_lock", "synchronize"),
            ("try_acquire", "synchronize"),
        ];
        for (compound, intent) in &compound_verbs {
            if two_word == *compound {
                let object = words[2..].join(" ");
                return NameSemantics {
                    verb: two_word,
                    object,
                    intent: intent.to_string(),
                };
            }
        }
    }

    // Fallback: the entire name is the verb
    NameSemantics {
        verb: name.to_string(),
        object: String::new(),
        intent: "unknown".to_string(),
    }
}

/// Categorize a verb into an intent category.
///
/// @docspec:deterministic
/// @docspec:intentional "Maps a verb string to one of 20+ intent categories using exhaustive pattern matching"
fn categorize_intent(verb: &str) -> String {
    match verb {
        "get" | "find" | "fetch" | "load" | "is" | "has" | "list" | "read" | "count"
            => "query",
        "set" | "update" | "save" | "add" | "write" | "flush"
            => "mutation",
        "create" | "build" | "generate" | "initialize" | "init" | "open" | "spawn"
            => "creation",
        "delete" | "remove" | "close" | "stop" | "drain"
            => "deletion",
        "validate" | "check" | "verify"
            => "validation",
        "compute" | "calculate" | "process"
            => "computation",
        "convert" | "parse" | "format" | "encode" | "decode" | "serialize" | "deserialize"
            => "transformation",
        "handle" | "dispatch" | "run" | "execute" | "start"
            => "handler",
        "send" | "emit"
            => "emission",
        "receive" | "poll"
            => "consumption",
        "sync"
            => "synchronize",
        "migrate"
            => "migrate",
        "schedule"
            => "schedule",
        "retry" | "try"
            => "retry",
        "batch"
            => "batch-process",
        "aggregate"
            => "aggregate",
        "merge"
            => "merge",
        "split"
            => "split",
        "notify"
            => "notify",
        "publish"
            => "publish",
        "subscribe"
            => "subscribe",
        "transform" | "enrich"
            => "transform",
        "filter"
            => "filter",
        "connect" | "bind" | "register" | "mount"
            => "lifecycle",
        "disconnect" | "unbind" | "unregister" | "unmount"
            => "lifecycle",
        "lock" | "acquire"
            => "synchronize",
        "unlock" | "release"
            => "synchronize",
        _
            => "unknown",
    }.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple_verb() {
        let result = analyze("get_user");
        assert_eq!(result.verb, "get");
        assert_eq!(result.object, "user");
        assert_eq!(result.intent, "query");
    }

    #[test]
    fn test_multi_word_object() {
        let result = analyze("create_payment_order");
        assert_eq!(result.verb, "create");
        assert_eq!(result.object, "payment order");
        assert_eq!(result.intent, "creation");
    }

    #[test]
    fn test_validation_verb() {
        let result = analyze("validate_input");
        assert_eq!(result.verb, "validate");
        assert_eq!(result.object, "input");
        assert_eq!(result.intent, "validation");
    }

    #[test]
    fn test_unknown_verb() {
        let result = analyze("frobnicate");
        assert_eq!(result.verb, "frobnicate");
        assert_eq!(result.object, "");
        assert_eq!(result.intent, "unknown");
    }

    #[test]
    fn test_underscore_prefix_stripped() {
        let result = analyze("_get_internal");
        assert_eq!(result.verb, "get");
        assert_eq!(result.object, "internal");
        assert_eq!(result.intent, "query");
    }

    #[test]
    fn test_is_check() {
        let result = analyze("is_valid");
        assert_eq!(result.verb, "is");
        assert_eq!(result.intent, "query");
    }

    #[test]
    fn test_delete_verb() {
        let result = analyze("delete_old_records");
        assert_eq!(result.verb, "delete");
        assert_eq!(result.object, "old records");
        assert_eq!(result.intent, "deletion");
    }
}
